### ISSUE TITLE
DiscriminatorMap is breaking Facet Aggregation Builders

### ISSUE THREAD

**Issue by Ludo444** [doctrine/mongodb-odm#2157]
### Bug Report

|    Q        |   A
|------------ | ------
| BC Break    | no
| Version     | 2.0.2

#### Summary

If there is a `Doctrine\ODM\MongoDB\Aggregation\Stage\Facet` used on a pipeline with document using `DiscriminatorMap`, when calling `getPipeline()` method inside `Facet`, it prepends the pipeline with filtering `$match` step that queries `DiscriminatorField`. This filtering may fail in certain pipelines as the field may be missing.

#### Current behavior

During later stages of a pipeline the `DiscriminatorField` may not be present in document list and `Facet` stage can contain empty results because of the filtering.

#### How to reproduce

Have a document that contains similar annotation:
```
 * @ODM\DiscriminatorField("type")
 * @ODM\DiscriminatorMap({"typ1"=Type1Document::class, "bracket"=Type2Document::class})
 */
abstract class Document
```

Use a `Facet` stage on an aggregation `Builder` that queries above type of `Document`.

#### Suggested solution

I think in case when `Builder` is used inside the `Facet` stage, it should ignore `DiscriminatorMap` filtering as the `DiscriminatorField` may not be present in pipeline anymore. Either by passing a param to `getPipeline()` method or let the `Builder` instance somehow know it should not add those extra filters there.

**Comment by malarzm** [comment]
@Ludo444 thanks for the report! Could you maybe create a PR with a failing test case? If you don't feel like diving too much into our tests, a functional one [like one of these](https://github.com/doctrine/mongodb-odm/tree/master/tests/Doctrine/ODM/MongoDB/Tests/Functional/Ticket) will suffice.

**Comment by Ludo444** [comment]
@malarzm hopefully i can look to it sometime this week

**Comment by Ludo444** [comment]
I created PR with test case and possible fix. Not sure if it's the most elegant solution but it seems to solve the problem fine.

**Comment by malarzm** [comment]
@Ludo444 thank you very much! I've asked Andreas to take a look there :)

**Comment by alcaeus** [comment]
Fixed in #2160. Thanks for the PR, @Ludo444 and sorry for the delay!
